### ISSUE TITLE
[FEATURE] WontPredict: meta model. 

### ISSUE THREAD

**Issue by koaning** [koaning/scikit-lego#235]
In the world of hype @MBrouns and myself came up with a very normal thought. 

<img width="611" alt="Screenshot 2019-10-28 at 20 00 04" src="https://user-images.githubusercontent.com/1019791/67709060-8f4f8800-f9bd-11e9-8ac2-a4566754913b.png">

One way to acomplish this is by introduction of a new meta model: `WontPredict`. 

This thread is meant as a place to discuss the implementation.

**Comment by koaning** [comment]
Here's my first proposal.

<img width="676" alt="Screenshot 2019-10-28 at 19 56 43" src="https://user-images.githubusercontent.com/1019791/67709212-c7ef6180-f9bd-11e9-8d59-47beec338ec3.png">

The idea is that there are two objects being passed. 

1. The original estimator that tries to predict `y` using `X` 
2. Another object, most likely an outlier detector, that determines if the current row needs to be avoided for prediction. 

### Possible Judges 

- based on outlier detection?
- based on classifier model? 
- based on proba confidence? 
- based on rule based system?

**Comment by MBrouns** [comment]
Some of the things that come to mind w.r.t. the implementation:

How are scorers going to account for the wontpredict cases. Should they simply ignore them? If so, how are we going to facilitate the user.

How are we going to indicate that a specific example is a wontpredict example?

I have a feeling both of these issues could be elegantly solved using numpys `maskedarray` implementation, but I'm not 100% sure.

**Comment by MBrouns** [comment]
Also, should we distinguish between probabilistic classifiers and outlierdetectors and their nonprobabilistic counterparts? If both are probabilistic it feels right to do `P(Class) = P(Class | Data) * P(Data)` and then thresholding the outcome of that to get binary predictions

**Comment by koaning** [comment]
In terms of scorers: 

> yeah, that stuff is gonna be a bit tricky. but this depends on what we use for a `WontPredict` flag. something suggests to me that `np.nan` is more general than `-1` because I'd also want this to work for regression models 

In terms of probabilistc counterparts:

> Regression needs to work too. Probably best to ignore the probabilistic classifiers as a special case for now. We could consider this to be a separate Meta Model (one with merit to it too).

**Comment by MBrouns** [comment]
Good point on sticking with the simple cases for now and treating `ProbabilisticWontPredict` separately. np.nan definitely seems more generic. i'll see if I have some time tomorrow to play around with maskedarray to get a better understanding of how that works.

**Comment by corriebar** [comment]
Regarding how to indicate that an example is a WontPredict, there are two clustering methods in sklearn (that I'm aware of), [DBSCAN](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html) and [OPTICS](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.OPTICS.html#sklearn.cluster.OPTICS), that output -1 as a label for noisy examples. Intuitively, this seems to me like a good approach to label WontPredicts and it also would be consistent with these methods. 
But indeed, wouldn't work with regression.

**Comment by MBrouns** [comment]
-1 works in part because both of those methods are clustering methods and therefore make their own cluster labels anyways. There's nothing stopping a regular classifier to have `-1` as a regular class although that use case might be rare.

**Comment by koaning** [comment]
Technically .. we could allow the user to set the flag. I imagine that in general the value for `wont-predict` is -1 but in the case of regression the user could specify `np.nan`? This could be passed upon initialising `WontPredict()`.

**Comment by koaning** [comment]
I've been mulling over this for a while and I am now sort of wondering if this is something we really want to automate in a `Pipeline`. 

If I were to look at the proposal from my talk; 

![image](https://user-images.githubusercontent.com/1019791/68526092-8ffaef00-02d8-11ea-94c0-97b4539dbf43.png)

This example feels like it is a pipeline that goes through stages. But another way of looking at this is: it is a DAG. Maybe that is something that is best handled with custom code, outside of the scope of scikit learn. @MBrouns any thoughts on this?

**Comment by koaning** [comment]
Alternatively, this might be an idea too. 

```
(CaseWhenEstimator()
  .when(predicate1, est1)
  .when(predicate2, est2)
  .when(predicate3, lambda X: -1)
  .default(then))
```

**Comment by MBrouns** [comment]
I kinda like this! as long as those predicates are estimators themselves
you can basically build a dag with multiple exit points.

Then again using this to build complex graphs is a bit iffy so we might
want to think about an alternative pipeline implementation that is a dag
rather than a pipeline.

**Comment by koaning** [comment]
@MBrouns here's a proposal! 

```
WontPredict(model_pipeline, outlier_pipeline) 
```

Here's the idea. If you run; 

```
model_pipeline.predict(x_new)
```

And theres an outlier ... then ... we just raise ...

```
WontPredictError()
```

This allows the user to catch it in production and handle it the way they like. Sound good?

**Comment by koaning** [comment]
I'll be closing this issue. 

If this idea is going to have merit then probably it is best to make a seperate package for it.
